code/modules/mob/living/carbon/human/ai/brain/ai_brain_cover.dm

```diff
@@ -2,40 +2,40 @@
 	/// If TRUE, AI is currently in some form of cover
 	var/in_cover = FALSE
 
-	/// If TRUE, the AI is in hard cover, meaning that it's solid and can't be seen through
-	var/hard_cover = FALSE
+	/// Reference to currently selected cover atom
+	var/atom/current_cover
 
 	COOLDOWN_DECLARE(cover_search_cooldown)
 
-/datum/human_ai_brain/proc/on_cover_destroyed(atom/source, force)
-	UnregisterSignal(source, COMSIG_PARENT_QDELETING)
+/datum/human_ai_brain/proc/end_cover(atom/source) // 'source' argument reserved for unregistering cade/wall/etc delete comsigs
+	current_cover = null
 	in_cover = FALSE
-	hard_cover = FALSE
-
-/datum/human_ai_brain/proc/on_shot_inside_cover(datum/source, damage_result, ammo_flags, obj/projectile/bullet)
-	if(faction_check(bullet.firer))
-		return // FF
 
+/datum/human_ai_brain/proc/on_shot_inside_cover(obj/projectile/bullet)
 	// Cover isn't working. Charge!
-	in_cover = FALSE
-	hard_cover = FALSE
-	UnregisterSignal(tied_human, COMSIG_HUMAN_BULLET_ACT)
+	end_cover()
 
 
-/datum/human_ai_brain/proc/locate_cover(obj/projectile/bullet, projectile_dir)
+/datum/human_ai_brain/proc/try_cover(obj/projectile/bullet)
 	if(!COOLDOWN_FINISHED(src, cover_search_cooldown))
 		return
+
 	COOLDOWN_START(src, cover_search_cooldown, 15 SECONDS)
+
 	var/list/turf_dict = list()
-	if(!recursive_turf_cover_scan(get_turf(tied_human), turf_dict, reverse_direction(projectile_dir)))
-		outside_cover_processing(bullet, projectile_dir, FALSE)
-		squad_cover_processing(FALSE, turf_dict, bullet, projectile_dir)
-	else
-		inside_cover_processing(turf_dict, FALSE)
-		squad_cover_processing(TRUE, turf_dict, bullet, projectile_dir)
+	var/cover_dir = reverse_direction(angle2dir4ai(bullet.angle))
 
+	recursive_turf_cover_scan(get_turf(tied_human), turf_dict, cover_dir)
```

Comment:
the reason the old system existed was that AI couldn't take cover anywhere that was considered "outdoors" because the tile count would exceed 200. Are you sure this works well outdoors

Reply:
I tried it multiple times while testing things on the LV624 and in GArrow hangar, and it worked pretty decent.
Considering i made it look for cover in random directions, only sometimes AI can't find a good spot to cover from enemy fire, but even in those moments it still looks like running in some direction which is pretty normal thing for combat. 
We might this restriction in future, in case some strange behavior will show up on live tests (again, local ones were fine)